### DISCUSSION TITLE
[schema] Cleanup infer/dontInfer

### DISCUSSION THREAD

**Discussion by freiksenet** [gatsbyjs/gatsby#18873] [Ideas / Feature Requests]
Currently we have non-symmetrical relation between directives `@infer`/`@dontInfer` and extensions `infer`. `@infer` maps to `infer: true`, while `@dontInfer` maps to `infer: false`. This not only breaks the convention that directives directly map onto extensions, but also the convention that extension values are objects.

I propose the following changes:

1. Add a parameter to `@infer` - `mode`. It accepts an enum of `ENABLED` | `DISABLED` | `DEPRECATED`
2. `@infer` default becomes  `@infer(mode: 'ENABLED')` and `{ infer: { mode: 'ENABLED' } }`.
3. `@dontInfer` stays the same and becomes `{ infer: { mode: 'DISABLED` } }`.
4. `infer: true` and `infer: false` get converted to corresponding object and trigger a deprecation warning.
5. `DEPRECATED` for infer raises a warning if declared schema doesn't match inferred schema. It is to be used, for example, by core team when we define core types like `File` or `ImageSharp`.

CC @vladar @stefanprobst @ChristopherBiscardi

**Comment by pieh** [comment]
Question about:
> `infer: true` and `infer: false` get converted to corresponding object and trigger a deprecation warning.

This is for when users/plugins declare type using `schema.buildObjectType`? If so, should using `@infer` (without params) and `@dontInfer` directives using SDL also trigger deprecation warning?

**Comment by freiksenet** [comment]
> This is for when users/plugins declare type using schema.buildObjectType?

Yes, they should do `infer: {}` or `infer: { mode: 'ENABLED' }`.

>  If so, should using `@infer` (without params) and `@dontInfer` directives using SDL also trigger deprecation warning?

I think it's fine to have defaults, so `@infer` is `infer: {}` that is a valid config.
